File: packages/tf-lang-l0-ts/src/ops/bounds.ts
interface BoundsOpts { min?: number; max?: number; inclusive?: boolean }

export function assert_bounds(v: unknown, opts: BoundsOpts): null {
  if (typeof v !== 'number') throw new Error('E_BOUNDS_TYPE');
  if (!Number.isInteger(v)) throw new Error('E_L0_FLOAT');
  const { min = Number.MIN_SAFE_INTEGER, max = Number.MAX_SAFE_INTEGER, inclusive = true } = opts || {};
  if (!Number.isInteger(min) || !Number.isInteger(max)) throw new Error('E_L0_FLOAT');
  if (inclusive) {
    if (v < min || v > max) throw new Error(`E_BOUNDS:${v}`);
  } else {
    if (v <= min || v >= max) throw new Error(`E_BOUNDS:${v}`);
  }
  return null;
}

File: packages/tf-lang-l0-rs/src/ops/bounds.rs
use anyhow::{bail, Result};
use serde_json::Value;

pub fn assert_bounds(v: &Value, opts: &Value) -> Result<Value> {
    let n = v.as_i64().ok_or_else(|| anyhow::anyhow!("E_BOUNDS_TYPE"))?;
    let min = opts.get("min").and_then(|x| x.as_i64()).unwrap_or(i64::MIN);
    let max = opts.get("max").and_then(|x| x.as_i64()).unwrap_or(i64::MAX);
    let inclusive = opts.get("inclusive").and_then(|x| x.as_bool()).unwrap_or(true);
    if inclusive {
        if n < min || n > max {
            bail!("E_BOUNDS:{}", n);
        }
    } else {
        if n <= min || n >= max {
            bail!("E_BOUNDS:{}", n);
        }
    }
    Ok(Value::Null)
}
